File: backend/QuanLyDiem/QuanLyDiem.API/Models/CourseClass.cs
using QuanLyDiem.API.Models;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace QuanLyDiem.API.Models
{
    public class CourseClass
    {
        [Key]
        public int CourseClassId { get; set; }

        [Required(ErrorMessage = "Mã lớp học phần không được để trống")]
        [MaxLength(50)]
        public string ClassCode { get; set; } // VD: IT01_N01

        [Required(ErrorMessage = "Vui lòng chọn học kỳ.")]
        public int SemesterId { get; set; }

        [ForeignKey("SemesterId")]
        public Semester? Semester { get; set; }

        [Required(ErrorMessage = "Vui lòng chọn môn học.")]
        public int SubjectId { get; set; }
        [ForeignKey("SubjectId")]
        public Subject? Subject { get; set; }

        public int? LecturerId { get; set; }
        [ForeignKey("LecturerId")]
        public User? Lecturer { get; set; }

        public ICollection<Enrollment>? Enrollments { get; set; } = new List<Enrollment>();
    }
}

File: frontend/client-app/src/app/services/course-class.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, map } from 'rxjs';

import { CourseClass } from '../models/course-class';

@Injectable({
  providedIn: 'root'
})
export class CourseClassService {

  private apiUrl = 'http://localhost:5059/api/CourseClasses';

  constructor(private http: HttpClient) {}

  getCourseClasses(): Observable<CourseClass[]> {
    return this.http.get<any[]>(this.apiUrl).pipe(
      map(data => data.map(item => ({
        id: item.courseClassId,
        classCode: item.classCode,
        subjectId: item.subjectId,
        semesterId: item.semesterId,
        lecturerId: item.lecturerId,
        subjectName: item.subject?.subjectName ?? '',
        lecturerName: item.lecturer?.fullName ?? 'Chưa phân công',
        semester: item.semester?.term ?? '',
        academicYear: item.semester?.academicYear ?? '',
        maxStudents: 40
      })))
    );
  }

  getCourseClassById(id: number): Observable<CourseClass> {
    return this.http.get<any>(`${this.apiUrl}/${id}`).pipe(
      map(item => ({
        id: item.courseClassId,
        classCode: item.classCode,
        subjectId: item.subjectId,
        semesterId: item.semesterId,
        lecturerId: item.lecturerId,
        subjectName: item.subject?.subjectName ?? '',
        lecturerName: item.lecturer?.fullName ?? 'Chưa phân công',
        semester: item.semester?.term ?? '',
        academicYear: item.semester?.academicYear ?? '',
        maxStudents: 40
      }))
    );
  }

  addCourseClass(courseClass: CourseClass): Observable<CourseClass> {
    return this.http.post<CourseClass>(this.apiUrl, {
      classCode: courseClass.classCode,
      subjectId: courseClass.subjectId,
      semesterId: courseClass.semesterId,
      lecturerId: courseClass.lecturerId ?? null
    });
  }

  updateCourseClass(courseClass: CourseClass): Observable<void> {
    return this.http.put<void>(
      `${this.apiUrl}/${courseClass.id}`,
      {
        courseClassId: courseClass.id,
        classCode: courseClass.classCode,
        subjectId: courseClass.subjectId,
        semesterId: courseClass.semesterId,
        lecturerId: courseClass.lecturerId ?? null
      }
    );
  }

  deleteCourseClass(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}